File: algorithms-week5/primitive_calculator/python/primitive_calculator.py
# Uses python3
import sys

def optimal_sequence(n):
    sequence = []
    ops = [0, 0]
    denominator = [2, 3]

    for i in range(2, n+1):
        ops.append(ops[i-1] + 1)

        for div in list(denominator):
            if (i % div) == 0:
                min = ops[int(i / div)]
                if min < ops[i]:
                    ops[i] = min + 1

    sequence.append(n)
    while n > 1:
        min_op = ops[n-1]
        m = n-1

        for div in list(denominator):
            if (n % div) == 0:
                op = ops[int(n / div)]
                if op < min_op:
                    min_op = op
                    m = int(n / div)

        sequence.append(m)
        n = m

    return reversed(sequence)

input = sys.stdin.read()
n = int(input)
sequence = list(optimal_sequence(n))
print(len(sequence) - 1)
for x in sequence:
    print(x, end=' ')

File: algorithms-week5/primitive_calculator/cpp/primitive_calculator.cpp
//
// Created by Sam on 8/31/2018.
//

#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;

vector<int> optimal_sequence(int n) {
  vector<int> ops = {0, 0};
  vector<int> divider = {2, 3};

  for (int i = 2; i <= n; i++) {
    ops.push_back(ops[i-1] + 1);

    for (const int& div : divider) {
      if ((i % div) == 0) {
        int min = ops[i / div];
        if (min < ops[i]) {
          ops[i] = min + 1;
        }
      }
    }
  }

  vector<int> sequence;
  sequence.push_back(n);
  while (n > 1) {
    int min_op = ops[n-1];
    int m = n - 1;

    for (const int& div : divider) {
      if ((n % div) == 0) {
        int op = ops[n / div];
        if (op < min_op) {
          min_op = op;
          m = n / div;
        }
      }
    }

    sequence.push_back(m);
    n = m;
  }
  reverse(sequence.begin(), sequence.end());
  return sequence;
}

int main() {
  int n;
  std::cin >> n;
  vector<int> sequence = optimal_sequence(n);
  std::cout << sequence.size() - 1 << std::endl;
  for (int i : sequence) {
    std::cout << i << " ";
  }
}
